### PR TITLE
Add deviation check

### PR DESCRIPTION
**Pull request by vponline**
This adds the functionality to check the DapiServer value for a **beaconId** and check if the updated API value deviation passes the deviationThreshold.

### CODE REVIEW

File: test/check-condition.feature.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import fs from 'fs';
+import path from 'path';
+import { ethers } from 'ethers';
+import { checkUpdateCondition } from '../src/check-condition';
+
+const dapiServerAbi = JSON.parse(fs.readFileSync(path.resolve('./src/test/artifacts/DapiServer.json')).toString());
```
Comment (vponline) [inline test/check-condition.feature.ts]: I used these tests to check that this actually worked with the Airkeeper DapiServer local PSP setup, and these could be used for E2E tests in the future. Otherwise I can delete if these are not needed.

File: src/check-condition.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
+  const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
+
+  // TODO readDataFeedValueWithId can throw if the dataFeed is not registered yet
+  // what do we want to return in this case
+  const dapiServerResponse = await dapiServer.connect(voidSigner).readDataFeedValueWithId(beaconId);
+  const updateInPercentage = calculateUpdateInPercentage(dapiServerResponse, ethers.BigNumber.from(updatedApiValue));
+
+  return updateInPercentage > deviationThreshold;
+};
```
Comment (vponline) [inline src/check-condition.ts]: Do we want a `try-catch` here for `readDataFeedValueWithId`?

Comment (bbenligiray) [reply src/check-condition.ts]: I assume you need to read the full data feed including the timestamp in Airseeker because it will need to update based on timestamp whatever the value is

Comment (vponline) [reply src/check-condition.ts]: :+1: Fixed. 

In this case, should I create a new function to read the value if it's needed in other places as well? Or will the timestamp logic be added here?

Comment (bbenligiray) [reply src/check-condition.ts]: We shouldn't do multiple calls to the blockchain provider to read the same value. @amarthadan can comment on how that should be implemented.

File: src/check-condition.test.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    expect(updateInPercentage).toEqual(80);
+  });
+});
+
+describe('checkUpdateCondition', () => {
+  const providerUrl = 'http://127.0.0.1:8545/';
+  const beaconId = '0x2ba0526238b0f2671b7981fd7a263730619c8e849a528088fd4a92350a8c2f2c';
+
+  let readDataFeedWithIdSpy: any;
+  let dapiServerMock: any;
```
Comment (acenolaza) [inline src/check-condition.test.ts]: Should we try to set proper types here?

File: src/check-condition.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+export const checkUpdateCondition = async (
+  providerUrl: string,
+  dapiServer: ethers.Contract,
+  beaconId: string,
+  deviationThreshold: number,
+  updatedApiValue: number
+) => {
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
+  const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
```
Comment (acenolaza) [inline src/check-condition.ts]: These 2 seems like they should be initialized only once instead of doing it for every condition check.

File: src/check-condition.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  // Convert back to a percentage number
+  return (change.toNumber() / HUNDRED_PERCENT) * 100;
+};
+
+export const checkUpdateCondition = async (
+  providerUrl: string,
+  dapiServer: ethers.Contract,
+  beaconId: string,
+  deviationThreshold: number,
+  updatedApiValue: number
```
Comment (acenolaza) [inline src/check-condition.ts]: Nit: just apiValue maybe? or mostUpToDateApiValue? or latestApiValue? or just keep current 😁

File: src/check-condition.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+) => {
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
+  const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
+
+  const dapiServerResponse = await dapiServer.connect(voidSigner).readDataFeedWithId(beaconId);
```
Comment (acenolaza) [inline src/check-condition.ts]: Maybe here we might want to just return `true` if current datafeed timestamp value is older than x amount of time? WDYT @bbenligiray and skip the condition check?

**Review by acenolaza — COMMENTED** [review]
A few comments but overall is looking good 👍🏻

File: src/check-condition.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  const delta = updatedValue.sub(initialValue);
+  const absoluteDelta = delta.abs();
+
+  // Avoid division by 0
+  const absoluteInitialValue = initialValue.isZero() ? ethers.BigNumber.from(1) : initialValue.abs();
+
+  const change = absoluteDelta.mul(ethers.BigNumber.from(HUNDRED_PERCENT)).div(absoluteInitialValue);
+
+  // Convert back to a percentage number
+  return (change.toNumber() / HUNDRED_PERCENT) * 100;
```
Comment (acenolaza) [inline src/check-condition.ts]: I have a really hard time trusting integers in js and I wonder if we should just keep doing the math using BigNumber instead 🤔

Comment (vponline) [reply src/check-condition.ts]: That makes sense to me now that you mentioned :smile:

File: src/check-condition.ts (hunk 2 above)
Comment (amarthadan) [reply src/check-condition.ts]: @bbenligiray Do you mean if different sponsors (on the same chain) want to read the same beacon? I'm not sure how that can be easily done since you can configure different update intervals for each sponsor. Or am I missing something?

Or do you mean the same chain, the same sponsor, and multiple providers?

File: src/check-condition.ts (hunk 4 above)
Comment (amarthadan) [reply src/check-condition.ts]: Yes, what's @acenolaza saying is true. You can expect the provider (an initialized provider object) to be passed to the function.

File: src/check-condition.ts (hunk 6 above)
Comment (amarthadan) [reply src/check-condition.ts]: That's going to be added as part of [BEC-297](https://api3dao.atlassian.net/browse/BEC-297).

File: src/check-condition.ts (hunk 4 above)
Comment (vponline) [reply src/check-condition.ts]: Then we could just pass in the voidSigner right? We don't need the provider itself.

File: src/check-condition.test.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
+import { ethers } from 'ethers';
+import { calculateUpdateInPercentage, checkUpdateCondition, HUNDRED_PERCENT } from './check-condition';
+
+describe('calculateUpdateInPercentage', () => {
```
Comment (Siegrift) [inline src/check-condition.test.ts:4]: I don't think we need so many tests. I'd drop the `calculates increase` and  `calculates decrease` since they are implicitely tested by the other ones.

File: src/check-condition.test.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
+
+  let readDataFeedWithIdSpy: jest.Mock;
+  let dapiServerMock: {
+    connect(_signerOrProvider: ethers.Signer | ethers.providers.Provider | string): typeof dapiServerMock;
+    readDataFeedWithId: jest.Mock;
+  };
+
+  beforeEach(() => {
+    const readDataFeedWithIdMock = (_beaconId: string) => Promise.resolve([ethers.BigNumber.from(500)]);
```
Comment (Siegrift) [inline src/check-condition.test.ts]: Nit: you don't need to specify arguments if you don't need to use them.

File: src/check-condition.test.ts [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  });
+
+  it('reads dapiserver value and checks the threshold condition to be true for decrease', async () => {
+    const readDataFeedWithIdOnceSpy = jest
+      .fn()
+      .mockImplementationOnce(() => Promise.resolve([ethers.BigNumber.from(400)]));
+    const checkUpdateConditionResult = await checkUpdateCondition(
+      voidSigner,
+      {
+        ...dapiServerMock,
```
Comment (Siegrift) [inline src/check-condition.test.ts]: Why can't you do the same mocking as in the beforeEach hook? I think it would be a bit cleaner not to use the beforeEach and mock `dapiServerMock` in each of the tests.

File: src/check-condition.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+// `calculateUpdateInPercentage()`. Since the reported data needs to fit
+// into 224 bits, its multiplication by 10^8 is guaranteed not to overflow.
+export const HUNDRED_PERCENT = 1e8;
+
+export const calculateUpdateInPercentage = (initialValue: ethers.BigNumber, updatedValue: ethers.BigNumber) => {
+  const delta = updatedValue.sub(initialValue);
+  const absoluteDelta = delta.abs();
+
+  // Avoid division by 0
+  const absoluteInitialValue = initialValue.isZero() ? ethers.BigNumber.from(1) : initialValue.abs();
```
Comment (Siegrift) [inline src/check-condition.ts:13]: Mhh, I wonder if this edge case should be explained somewhere (for the API integrators).

File: src/check-condition.ts [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export const checkUpdateCondition = async (
+  voidSigner: ethers.VoidSigner,
+  dapiServer: ethers.Contract,
+  beaconId: string,
+  deviationThreshold: number,
+  apiValue: number
+) => {
+  const [dapiServerValue, _timestamp] = await dapiServer.connect(voidSigner).readDataFeedWithId(beaconId);
```
Comment (Siegrift) [inline src/check-condition.ts:25]: We also want to to update if the response is too old (not sure if that is part of this task though)

File: test/check-condition.feature.ts [hunk 13]
```diff
@@ -0,0 +1,83 @@
+import { ethers } from 'ethers';
+import { DapiServer__factory as DapiServerFactory } from '@api3/airnode-protocol-v1';
+import { checkUpdateCondition } from '../src/check-condition';
+
+describe('checkUpdateCondition', () => {
+  const providerUrl = 'http://127.0.0.1:8545/';
+  const provider = new ethers.providers.JsonRpcProvider(providerUrl);
+  const voidSigner = new ethers.VoidSigner(ethers.constants.AddressZero, provider);
+  const dapiServer = new ethers.Contract('0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0', DapiServerFactory.abi, provider);
```
Comment (Siegrift) [inline test/check-condition.feature.ts]: 1) I am missing scripts to deploy this contract
2) This test is not seem to be recognized by jest configuration.
3) Can you rename the `e2e.test.ts` to `make-request.feature.ts`? Also, maybe put them into folder `e2e`.

**Review by Siegrift — CHANGES_REQUESTED** [review]

File: test/check-condition.feature.ts (hunk 13 above)
Comment (vponline) [reply test/check-condition.feature.ts]: I will cover all of these in the Airseeker E2E task.

(if you wanted to run these you could deploy the Airkeeper local PSP + run the update script).

File: src/check-condition.ts (hunk 12 above)
Comment (vponline) [reply src/check-condition.ts:25]: Will be covered in another task :smile:

File: src/check-condition.ts (hunk 11 above)
Comment (vponline) [reply src/check-condition.ts:13]: What kind of case were you thinking? (I thought this is just a math limitation)

**Review by acenolaza — APPROVED** [review]

File: src/check-condition.ts (hunk 11 above)
Comment (Siegrift) [reply src/check-condition.ts:13]: > (I thought this is just a math limitation)

It is, but we have to handle it somehow and using 1 seems the best we could do.

> What kind of case were you thinking?

If the value was `0` and now is `5` it doesn't make sense to ask "how big is the deviation" (because there is the division by 0). We divide by 1 in that case, which seems reasonable but it behaves strangely since deviation from 0 -> 1 leads the same percentage update as 1 -> 2.

File: test/check-condition.feature.ts (hunk 13 above)
Comment (Siegrift) [reply test/check-condition.feature.ts]: I think at least this test should be included in the pipeline before the PR is merged.

**Review by Siegrift — COMMENTED** [review]
👍 The functionality LGTM, but I think the e2e test should be included in the pipeline.

File: test/check-condition.feature.ts (hunk 13 above)
Comment (vponline) [reply test/check-condition.feature.ts]: Ok, I can add the setup for these from the E2E branch :smile:

File: src/check-condition.ts (hunk 11 above)
Comment (vponline) [reply src/check-condition.ts:13]: > We divide by 1 in that case, which seems reasonable but it behaves strangely since deviation from 0 -> 1 leads the same percentage update as 1 -> 2.

Shouldn't this be fine as the result is that the beacon should be updated? The actual update percentage doesn't really matter in this case (when the **dapiServerValue** is 0), right? Perhaps @amarthadan can comment as well.

(For context, this function is essentially the same as the smart contract function [here](https://github.com/api3dao/airnode/blob/v1-protocol/packages/airnode-protocol-v1/contracts/dapis/DapiServer.sol#L862))

Comment (Siegrift) [reply src/check-condition.ts:13]: Yeah, the value is 0 initially, so after the first beacon update should always update the value on chain. 
However, the value *could* be also set to 0 by the API and it's not clear whether the next non zero value should also trigger update. But this is an edge case, which I can't imagine happening in practice, so I think the implementation is good.

File: test/config/config.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
+export const buildAirseekerConfig = () => ({
```
Comment (Siegrift) [inline test/config/config.ts]: It would be better to move this to `test/fixtures`

File: test/setup/deployment.ts [hunk 15]
```diff
@@ -0,0 +1,231 @@
+import { Contract, Wallet } from 'ethers';
+import * as hre from 'hardhat';
+import * as abi from '@api3/airnode-abi';
+// TODO uncomment once airnode-node 0.6 is released
```
Comment (Siegrift) [inline test/setup/deployment.ts]: Nit: (more like my ocd :D)

```
TODO: uncomment...
```

File: tsconfig.json [hunk 16]
```diff
@@ -10,8 +10,8 @@
     "target": "es5",
     "skipLibCheck": true,
     "baseUrl": "./",
-    "outDir": "./dist/src"
+    "outDir": "./dist"
   },
-  "include": ["./src/**/*.ts"],
+  "include": ["./src/**/*.ts", "./test/**/*", "./hardhat.config.ts"],
```
Comment (Siegrift) [inline tsconfig.json]: This is OK for now, but we should have two separate tsconfigs (one for development and one for prod), but this is tracked in https://api3dao.atlassian.net/browse/BEC-311

**Review by Siegrift — APPROVED** [review]
👍 LGTM

File: .github/workflows/main.yml [hunk 17]
```diff
@@ -48,6 +48,26 @@ jobs:
...
+          cache: 'yarn'
+      - name: Install Dependencies
+        run: yarn install --frozen-lockfile
+      - name: Build
+        run: yarn build
+      - name: Start background services
+        run: yarn dev:testing-services:start
+      - name: Test
+        run: yarn test:e2e
+
```
Comment (amarthadan) [inline .github/workflows/main.yml]: Is this actually faster than having it run after the unit tests? If so, I'm ok with it but you need to add the `if` key to stop tests from running twice for PRs and you need to remove `yarn dev:testing-services:start` from the `lint-build-test` job above.

File: package.json [hunk 18]
```diff
@@ -12,8 +12,10 @@
     "build": "yarn clean && yarn compile",
     "clean": "rimraf -rf ./dist *.tgz",
     "compile": "tsc -p tsconfig.json",
-    "dev:api": "ts-node test/server.ts",
-    "dev:testing-services:start": "pm2 start testing-services.config.js",
+    "dev:clean": "rimraf -rf ./logs",
```
Comment (amarthadan) [inline package.json]: Why do we even create the `./logs` directory? If it's just for tests can't we log either to STDOUT or `/dev/null` or turn it off altogether if we don't need it?

File: src/check-condition.ts (hunk 11 above)
Comment (amarthadan) [reply src/check-condition.ts:13]: I agree with both of you. I think we should follow the implementation from the contract and I agree that this implementation is fine.

File: tsconfig.json (hunk 16 above)
Comment (amarthadan) [inline tsconfig.json]: Why do we need this? Tests are actually excluded in https://github.com/api3dao/airseeker/pull/28 to make the build process work correctly.

**Review by amarthadan — APPROVED** [review]
I'm approving this but please take a look at my comments before merging.

File: package.json (hunk 18 above)
Comment (vponline) [reply package.json]: These were removed from Airkeeper as well so I'll remove

File: .github/workflows/main.yml (hunk 17 above)
Comment (vponline) [reply .github/workflows/main.yml]: I dont believe it is faster, I'll change it to run after the unit tests (like you had done in beacon-updates)
